File: src/main/java/planetGenerator/ui/planet/viewProviders/BiomeViewProvider.java
package planetGenerator.ui.planet.viewProviders;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.layout.Pane;
import javafx.scene.paint.Color;
import planetGenerator.model.Biome;
import planetGenerator.model.Planet;
import planetGenerator.model.Region;
import rak.utility.MathFunctions;

public class BiomeViewProvider extends GraphicsContextViewProvider {
	private static final double DEFAULT_VARIATION = 100;
	private static double ALTITUDE_CLAMP = .5;
	private static double PERCIPITATION_CLAMP = .7;
	
	@Override
	public void drawView(Planet planet, Pane viewBackground) {
		GraphicsContext gg = createGraphicsContext(viewBackground, planet.getSize());
		for (int x=0; x<planet.getSize(); x++){
			for (int y=0; y<planet.getSize(); y++){
				Region region = planet.getRegion(x, y);
				Color color = getAdjustedColor(region);
				gg.setFill(color);
				gg.fillRect(x, y, 1, 1);
			}
		}
	}
	
	private Color getAdjustedColor(Region region){
		Biome biome = region.getBiome();
		double brightnessFactor = getFactor(region.getAltitude(), biome.getAltitude(), biome.getAltitudeVariation(), ALTITUDE_CLAMP);
		double saturationFactor = getFactor(region.getPercipitation(), biome.getPercipitation(), biome.getPercipitationVariation(), PERCIPITATION_CLAMP);
		double hueShift = 0;
		
		Color color = biome.getColor();
		return color.deriveColor(hueShift, saturationFactor, brightnessFactor, 1);
	}
	
	private double getFactor(int regionValue, int biomeValue, int variation, double clamp) {
		double biomeVariation = variation != 0 ? variation : DEFAULT_VARIATION;
		
		int regionVariation = regionValue - biomeValue;
		
		double percent = regionVariation/biomeVariation;
		percent = MathFunctions.clamp(percent, -clamp, clamp);
		
		return percent + 1;
	}
	
}

File: src/main/java/rak/planetGenerator/ui/planet/viewProviders/BiomeViewProvider.kt
package rak.planetGenerator.ui.planet.viewProviders

import javafx.scene.layout.Pane
import javafx.scene.paint.Color
import rak.planetGenerator.model.Planet
import rak.planetGenerator.model.Region
import rak.utility.MathFunctions

class BiomeViewProvider : GraphicsContextViewProvider() {
    private val DEFAULT_VARIATION: Int = 100
    private val ALTITUDE_CLAMP = .5
    private val PERCIPITATION_CLAMP = .7

    override fun drawView(planet: Planet, viewBackground: Pane) {
        val gg = createGraphicsContext(viewBackground, planet.size)
        for (x in 0 until planet.size) {
            for (y in 0 until planet.size) {
                val region = planet.getRegion(x, y)
                val color = getAdjustedColor(region!!)
                gg.fill = color
                gg.fillRect(x.toDouble(), y.toDouble(), 1.0, 1.0)
            }
        }
    }

    private fun getAdjustedColor(region: Region): Color {
        val biome = region.biome
        val brightnessFactor = getFactor(region.altitude, biome.altitude, biome.altitudeVariation, ALTITUDE_CLAMP)
        val saturationFactor = getFactor(region.percipitation, biome.percipitation, biome.percipitationVariation, PERCIPITATION_CLAMP)
        val hueShift = 0.0

        return biome.color.deriveColor(hueShift, saturationFactor, brightnessFactor, 1.0)
    }

    private fun getFactor(regionValue: Int, biomeValue: Int, variation: Int, clamp: Double): Double {
        val biomeVariation = if (variation != 0) variation else DEFAULT_VARIATION

        val regionVariation = regionValue - biomeValue

        var percent: Double = regionVariation / biomeVariation.toDouble()
        percent = MathFunctions.clamp(percent, -clamp, clamp)

        return percent + 1
    }

}
